fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java

```diff
@@ -0,0 +1,338 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+package com.starrocks.sql.parser;
+
+import com.google.common.collect.Lists;
+import com.starrocks.analysis.Expr;
+import com.starrocks.analysis.FunctionParams;
+import com.starrocks.analysis.GroupByClause;
+import com.starrocks.analysis.JoinOperator;
+import com.starrocks.analysis.LimitElement;
+import com.starrocks.analysis.OrderByElement;
+import com.starrocks.analysis.ParseNode;
+import com.starrocks.analysis.SelectList;
+import com.starrocks.analysis.SelectListItem;
+import com.starrocks.analysis.SlotRef;
+import com.starrocks.analysis.TableName;
+import com.starrocks.analysis.ValueList;
+import com.starrocks.sql.ast.CTERelation;
+import com.starrocks.sql.ast.ExceptRelation;
+import com.starrocks.sql.ast.Identifier;
+import com.starrocks.sql.ast.IntersectRelation;
+import com.starrocks.sql.ast.JoinRelation;
+import com.starrocks.sql.ast.QualifiedName;
+import com.starrocks.sql.ast.QueryRelation;
+import com.starrocks.sql.ast.QueryStatement;
+import com.starrocks.sql.ast.Relation;
+import com.starrocks.sql.ast.SelectRelation;
+import com.starrocks.sql.ast.TableFunctionRelation;
+import com.starrocks.sql.ast.TableRelation;
+import com.starrocks.sql.ast.UnionRelation;
+import com.starrocks.sql.ast.ValuesRelation;
+import com.starrocks.sql.optimizer.base.SetQualifier;
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.antlr.v4.runtime.tree.TerminalNode;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.toList;
+
+public class AstBuilder extends StarRocksBaseVisitor<ParseNode> {
+    @Override
+    public ParseNode visitSingleStatement(StarRocksParser.SingleStatementContext context) {
+        return visit(context.statement());
+    }
+
+    // -------------------------------- Statement ------------------------------
+
+    @Override
+    public ParseNode visitQueryStatement(StarRocksParser.QueryStatementContext context) {
+        QueryRelation queryRelation = (QueryRelation) visit(context.query());
+        return new QueryStatement(queryRelation);
+    }
+
+    @Override
+    public ParseNode visitQuery(StarRocksParser.QueryContext context) {
+        QueryRelation queryRelation = (QueryRelation) visit(context.queryNoWith());
+
+        List<CTERelation> withQuery = new ArrayList<>();
+        if (context.withClause() != null) {
+            withQuery = visit(context.withClause().commonTableExpression(), CTERelation.class);
+        }
+        withQuery.forEach(queryRelation::addCTERelation);
+
+        return queryRelation;
+    }
+
+    @Override
+    public ParseNode visitCommonTableExpression(StarRocksParser.CommonTableExpressionContext context) {
+        Optional<List<Identifier>> columns = Optional.empty();
```

Comment:
why not just use null list?

Reply:
Two style optional & nullable should be consistent, looks like most code follow the nullable style?

Reply:
done